module/zfs/zio.c

```diff
@@ -3316,18 +3337,60 @@ zio_ddt_free(zio_t *zio)
 	spa_t *spa = zio->io_spa;
 	blkptr_t *bp = zio->io_bp;
 	ddt_t *ddt = ddt_select(spa, bp);
-	ddt_entry_t *dde;
-	ddt_phys_t *ddp;
+	ddt_entry_t *dde = NULL;
+	ddt_phys_t *ddp = NULL;
+	boolean_t added = B_FALSE;
 
 	ASSERT(BP_GET_DEDUP(bp));
 	ASSERT(zio->io_child_type == ZIO_CHILD_LOGICAL);
 
 	ddt_enter(ddt);
-	freedde = dde = ddt_lookup(ddt, bp, B_TRUE);
+	freedde = dde = ddt_lookup(ddt, bp, B_FALSE, &added);
 	if (dde) {
 		ddp = ddt_phys_select(dde, bp);
-		if (ddp)
-			ddt_phys_decref(ddp);
+	}
+	if (ddp != NULL) {
+		ddt_phys_decref(ddp);
+	} else {
+		/*
+		 * ddt_phys_select() failed, because this entry was evicted
+		 * from the DDT (see ddt_sync_table()).  It can fail in two
```

Comment:
I don't see any changes to ddt_sync_table(), and I don't think entries can be evicted from the DDT.